lib/Value/Matrix.pm

```diff
@@ -493,6 +493,110 @@ sub I {
 	return $self->make($context, @M);
 }
 
+#
+#  Get an elementary matrix of the requested size and type
+#  E(n,[i,j])   nxn, swap rows i and j
+#  E(n,[i,j],k) nxn, replace row i with row i added to k times row j
+#  E(n,[i],k)   nxn, scale row i by k
+#
+sub E {
+	my ($self, $d, $rows, $k, $context) = @_;
+	$context = $self->context unless $context;
+	Value::Error("You must provide a dimension for an Elementary matrix") unless defined $d;
+	Value::Error("Dimension must be a positive integer")                  unless $d =~ m/^[1-9]\d*$/;
+	my @ij = @{$rows};
+	Value::Error("Either one or two rows must be specified for an Elementary matrix") unless (@ij == 1 || @ij == 2);
+	Value::Error(
+		"If only one row is specified for an Elementary matrix, then a number to scale by must also be specified")
+		if (@ij == 1 && !defined $k);
+	for (@ij) {
+		Value::Error("Row numbers must be integers between 1 and $d")
+			unless ($_ =~ m/^[1-9]\d*$/ && $_ >= 1 && $_ <= $d);
+	}
+	@ij = map { $_ - 1 } (@ij);
+
+	my @M    = ();
+	my @Z    = split('', 0 x $d);
+	my $REAL = $context->Package('Real');
+
+	foreach my $i (0 .. $d - 1) {
+		my @row = @Z;
+		$row[$i] = 1;
+		if (@ij == 1) {
+			$row[$i] = $k if ($i == $ij[0]);
+		} elsif (defined $k) {
+			$row[ $ij[1] ] = $k if ($i == $ij[0]);
+		} else {
+			($row[ $ij[0] ], $row[ $ij[1] ]) = ($row[ $ij[1] ], $row[ $ij[0] ]) if ($i == $ij[0] || $i == $ij[1]);
+		}
+		push(@M, $self->make($context, map { $REAL->new($_) } @row));
+	}
+	return $self->make($context, @M);
+}
+
+#
+#  Get a permutation matrix of the requested size
+#  E.g. P(3,[1,2,3])  corresponds to cycle (123) applied to rows of I_3
+#  And  P(6,[1,4],[2,4,6]) corresponds to cycle product (14)(246) applied to rows if I_6
+#
+sub P {
+	my ($self, $d, @cycles) = @_;
+	my $context = $self->context;
+	$d = ($self->dimensions)[0] if !defined $d && ref($self) && $self->isSquare;
+	Value::Error("You must provide a dimension for a Permutation matrix") unless defined $d;
+	Value::Error("Dimension must be a positive integer")                  unless $d =~ m/^[1-9]\d*$/;
+	for my $c (@cycles) {
+		Value::Error("Permutation cycles should be array references") unless (ref($c) eq 'ARRAY');
+		for (@$c) {
+			Value::Error("Permutation cycle indices must be integers between 1 and $d")
+				unless ($_ =~ m/^[1-9]\d*$/ && $_ >= 1 && $_ <= $d);
+		}
+		my %cycle_hash = map { $_ => '' } (@$c);
+		Value::Error("A permutation cycle should not repeat an index") unless (@$c == keys %cycle_hash);
+	}
+	my @M    = ();
+	my @Z    = split('', 0 x $d);
+	my $REAL = $context->Package('Real');
+
+	foreach my $i (0 .. $d - 1) {
+		my @row = @Z;
+		$row[$i] = 1;
+		push(@M, $self->make($context, map { $REAL->new($_) } @row));
+	}
+
+	for my $c (@cycles) {
+		my $n = @$c;
+		my $swap;
+		for my $i (0 .. $n - 1, 0) {
+			($swap, $M[ $c->[$i] - 1 ]) = ($M[ $c->[$i] - 1 ], $swap);
+		}
+	}
```

Comment:
This should be
```suggestion
	for my $c (@cycles) {
		my $swap;
		for my $i (0 .. $#$c, 0) {
			($swap, $M[ $c->[$i] - 1 ]) = ($M[ $c->[$i] - 1 ], $swap);
		}
	}
```
Perl already saves the index of the last entry in an array.  So just use that instead of computing it with `@$c - 1` and saving it in `$n`.